src/platform/Linux/MdnsImpl.cpp

```diff
@@ -540,25 +541,45 @@ CHIP_ERROR MdnsAvahi::Browse(const char * type, MdnsServiceProtocol protocol, ch
     return browser == nullptr ? CHIP_ERROR_INTERNAL : CHIP_NO_ERROR;
 }
 
-MdnsServiceProtocol TruncateProtocolInType(char * type)
+MdnsServiceProtocol GetProtocolInType(const char * type)
 {
-    char * deliminator           = strrchr(type, '.');
-    MdnsServiceProtocol protocol = MdnsServiceProtocol::kMdnsProtocolUnknown;
+    const char * deliminator = strrchr(type, '.');
 
-    if (deliminator != NULL)
+    if (deliminator == NULL)
     {
-        if (strcmp("._tcp", deliminator) == 0)
-        {
-            protocol     = MdnsServiceProtocol::kMdnsProtocolTcp;
-            *deliminator = 0;
-        }
-        else if (strcmp("._udp", deliminator) == 0)
-        {
-            protocol     = MdnsServiceProtocol::kMdnsProtocolUdp;
-            *deliminator = 0;
-        }
+        ChipLogError(Discovery, "Failed to find protocol in type: %s", type);
+        return MdnsServiceProtocol::kMdnsProtocolUnknown;
+    }
+
+    if (strcmp("._tcp", deliminator) == 0)
+    {
+        return MdnsServiceProtocol::kMdnsProtocolTcp;
+    }
+    if (strcmp("._udp", deliminator) == 0)
+    {
+        return MdnsServiceProtocol::kMdnsProtocolUdp;
+    }
+
+    ChipLogError(Discovery, "Unknown protocol in type: %s", type);
+    return MdnsServiceProtocol::kMdnsProtocolUnknown;
+}
+
+/// Copies the type from a string containing both type and protocol
+///
+/// e.g. if input is "foo.bar", output is "foo", input is 'a.b._tcp", oputput is "a.b"
+template <size_t N>
+void TypeFromTypeAndProtocol(char (&dest)[N], const char * typeAndProtocol)
+{
+    const char * dotPos          = strrchr(typeAndProtocol, '.');
```

Comment:
static_assert(N > 0)

Reply:
Is this needed? we are very unlikely to create and pass in arrays of size 0.

Reply:
Also assuming size 0 ... I think code will still be correct (but nothing useful will come out of it as even null terminator will be missing)